### ISSUE TITLE
Dynlink.loadfile_private behaves unexpectedly (?) on bytecode

### ISSUE THREAD

**Issue by glondu** [ocaml/ocaml#13374]
With OCaml 5.2.0:
```
$ cat a.ml
let x = "Hello!"
$ cat b.ml
let () = print_endline A.x
$ cat main.ml
let () = Dynlink.loadfile_private (Dynlink.adapt_filename "c.cma")
$ ocamlc -c a.ml
$ ocamlc -c b.ml
$ ocamlc -a -o c.cma a.cmo b.cmo
$ ocamlc -I +dynlink dynlink.cma main.ml
$ ./a.out
Fatal error: exception Dynlink.Error (Dynlink.Linking_error ("/tmp/ocaml/c.cma", Dynlink.Undefined_global "compilation unit `A'"))
```

I am expecting modules loaded by a plugin to be available to other modules loaded by the same plugin (which is the case in native code)... and I am not alone: Why3 seems to rely on this behaviour.

**Comment by lthls** [comment]
As a workaround, you should be able to pack the individual plugin units before making a library:
```shell-session
$ ocamlc -pack -o c.cmo a.cmo b.cmo
$ ocamlc -a -o c.cma c.cmo
```
I can reproduce the issue on 4.14 too, so it's not new. Did you get hit while upgrading to OCaml 5.x on architectures where native support was dropped ?

**Comment by glondu** [comment]
> I can reproduce the issue on 4.14 too, so it's not new. Did you get hit while upgrading to OCaml 5.x on architectures where native support was dropped ?

Yes. Actually, why3 tests have always been broken on bytecode, but with i386 becoming bytecode there is a new failure which is interpreted as a regression by our automatic QA.

As to whether this is a bug or feature: I do think that the behaviour (whichever is deemed correct) should be consistent between native code and bytecode.

**Comment by lthls** [comment]
Fixed on trunk by #13376.
